content/docs/troubleshooting/_index.md

```diff
@@ -280,3 +280,58 @@ field once the Ingress resource is ready to route traffic.
 For the Traefik controller, verify that the `kubernetes.ingressEndpoint` config
 is [set properly](https://docs.traefik.io/providers/kubernetes-ingress/). This option was
 introduced in Traefik 1.7.0.
+
+## Synchronous call was made to "X" with an unregistered Provider. {#synchronous-call}
+
+The warning occurs when making a 'data source' call in Pulumi in a synchronous fashion and passing along a ProivderResource that has not been registered. For example:
+
+```ts
+const provider = new aws.Provider(...);
+
+// A call to some provider's `getXXX` data source function.
+const ids = aws.ec2.getSubnetIds(..., { provider });
+
+// or
+
+const parent = new SomeResource("name", { provider });
+const ids = aws.ec2.getSubnetIds(..., { parent });
+```
+
+This warning may be benign. However, if you are experiencing crashes or hangs in Pulumi (especially in Node.js version 12.0 and above)
+and you see this warning, then it is likely that this is the source.
+
+Currently, a warning is issued so as to not break existing code that is functionality properly. In a future version, Pulumi *may* be
+updated to throw instead of producing a warning when this happens.  It is recommended that Pulumi apps be updated to prevent breakage
+in the future.
+
+To address the issue update your app to use one of the following forms:
+
+### Invoke the data-source function asynchronously:
+
+```ts
+const ids = pulumi.output(aws.ec2.getSubnetIds(..., { provider, async: true })); // or
+const ids = pulumi.output(aws.ec2.getSubnetIds(..., { parent, async: true }));
+```
+
+This is the preferred way to solve this issu. In this form, the `async: true` flag is passed in which forces `getSubnetIds` to always
+execute asynchronously.  Because of this, the result of the call is wrapped into an `Output` to allow it to be easily passed around and
+to make it [simple to access properties](https://www.pulumi.com/docs/intro/concepts/programming-model/#lifting) off of it.
+
+Sometimes, however, the above is not possible because the call the data-source happens a deeper layer (possibly in a component not under your control).  In that case:
+
+### Register the provider first:
+
+```ts
+const provider = new aws.Provider(...);
+await ProviderResource.register(provider);
+
+const ids = await aws.ec2.getSubnetIds(..., { provider }); // or
+const ids = await aws.ec2.getSubnetIds(..., { parent });
+```
+
+In this form, the ProviderResource is explicitly registered first, allowin it to be safely used *synchronously* in the data-source
+calls. This registration should generally be done right after creating the provider. With this form the data-source results can be used immediately, without needing to operate on them as promises (i.e. no need for `await` or
+`.then(...)`).
+
+This approach can always safely make it possible to perform these synchronous data-source calls.  However, it may come with significant
```

Comment:
```suggestion
This approach can always safely make it possible to perform these synchronous data-source calls.  However, it may require refactoring some code
```